toolkit/bridge/pallet/src/lib.rs

```diff
@@ -0,0 +1,236 @@
+//! Pallet that tracks information about incoming token bridge transfers observed on Cardano.
+#![cfg_attr(not(feature = "std"), no_std)]
+#![deny(missing_docs)]
+
+extern crate alloc;
+
+pub use pallet::*;
+
+/// Runtime logic for handling incoming token bridge transfers from Cardano
+///
+/// The chain builder should implement in accordance with their particular business rules and
+/// ledger structure. Calls to all functions defined by this trait should not return any errors
+/// as this would fail the block creation. Instead, any validation and business logic errors
+/// should be handled gracefully inside the handler code.
+pub trait TransferHandler<RecipientAddress, TokenAmount, SpecialTransferId> {
+	/// Should handle an incoming token transfer of `token_mount` tokens to `recipient`
+	fn handle_incoming_transfer(recipient: RecipientAddress, token_amount: TokenAmount);
+
+	/// Should handle an incoming special token transfer
+	///
+	/// Special transfers are transfers done as part of the protocol operation, as opposed to
+	/// user-initiated transfers, and can be defined by each Partner Chain according to its needs.
+	fn handle_incoming_special_transfer(id: SpecialTransferId, token_amount: TokenAmount);
+
+	/// Should handle invalid transfers
+	///
+	/// Invalid transfers come from malformed UTXOs on Cardano that do not contain data that can be
+	/// correctly interpreted. This function will be passed the total of tokens contained in all such
+	/// invalid utxos observed in the current block. If the Partner Chain does not implement any
+	/// mechanism for recovering funds from invalid transfers, this function can be left uninmplemented.
+	fn handle_invalid_transfers(_total_token_amount: TokenAmount) {}
```

Comment:
I see `handle_transfer` that has `target` parameter that is either RecipientAddress, ProtocolOperationId, or Malformed Data, instead of 3 functions.

Having one such type that can (but doesn't have to) contain both RecipientAddress and ProtocolOperationId variants will decrease number of (related) types in the pallet config.